src/funnels/Swapper.sol

```diff
@@ -0,0 +1,113 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.16;
+
+interface RolesLike {
+    function canCall(bytes32, address, address, bytes4) external view returns (bool);
+}
+
+interface GemLike {
+    function balanceOf(address) external view returns (uint256);
+    function transfer(address, uint256) external;
+    function transferFrom(address, address, uint256) external;
+}
+
+interface CalleeLike {
+    function swap(address, address, uint256, uint256, address, bytes calldata) external;
+}
+
+contract Swapper {
+    mapping (address => uint256) public wards;                         // Admins
+    mapping (address => mapping (address => PairLimit)) public limits; // Rate limit parameters per src->dst pair
+
+    RolesLike public immutable roles;  // Contract managing access control for this Swapper
+    bytes32   public immutable ilk;    // Collateral type
+    address   public immutable buffer; // Contract from which the GEM to sell is pulled and to which the bought GEM is pushed
+
+    struct PairLimit {
+        uint96  cap; // Maximum amount of src token that can be swapped each era for a src->dst pair
+        uint32  era; // Cooldown period it has to wait for renewing the due amount to cap for src to dst swap
+        uint96  due; // Pending amount of src token that can still be swapped until next era
+        uint32  end; // Timestamp of when the current batch ends
+    }
+
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event SetLimits(address indexed src, address indexed dst, uint96 cap, uint32 era);
+    event Swap(address indexed sender, address indexed src, address indexed dst, uint256 amt, uint256 out);
+
+    constructor(address roles_, bytes32 ilk_, address buffer_) {
+        roles = RolesLike(roles_);
+        ilk = ilk_;
+        buffer = buffer_;
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    modifier auth() {
+        require(roles.canCall(ilk, msg.sender, address(this), msg.sig) || wards[msg.sender] == 1, "Swapper/not-authorized");
+        _;
+    }
+
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function setLimits(address src, address dst, uint96 cap, uint32 era) external auth {
+        limits[src][dst] = PairLimit({
+            cap: cap,
+            era: era,
+            due: 0,
+            end: 0
+        });
+        emit SetLimits(src, dst, cap, era);
+    }
+
+    function swap(address src, address dst, uint256 amt, uint256 minOut, address callee, bytes calldata data) external auth returns (uint256 out) {
+        PairLimit memory limit = limits[src][dst];
+
+        if (block.timestamp >= limit.end) {
+            // Reset batch
+            limit.due = limit.cap;
+            limit.end = uint32(block.timestamp) + limit.era;
```

Comment:
So there is sort of a reentrancy issue if `limit.era` is zero--the full `cap` amount can be swapped for multiple times because `limit.due` will be reset upon re-entry (until the block gas limit is reached). I say "sort of" because that actually seems consistent with what the behavior should be if the time period is literally zero, in the strictest sense (zero cooldown so can just do another swap immediately...). But it might not be desirable/intended.

Separately, one can consider whether it would be better to express the period in terms of blocks rather than seconds. This is not a big deal if the typical `era` time used will be large compared to the block time (say, at least 10x). Times are also a little more natural to reason about. A chain stoppage scenario is usually the main concern that would motivate using blocks. Imagine: just after changing to a new `era`, the full `cap` is swapped and shortly after that the chain goes down for longer than the `era` time. The chain restarts, without DEX liquidity having been replenished or prices adjusted, but another swap can be done immediately because `block.timestamp` has advanced. In such a case, it might be preferable to have a block-based time unit.

Reply:
Regarding the reentrancy - in the special case where the rate limiting period `era` is set to 0, the rate limit should be considered disabled and the operator is technically able to drain the full `AllocatorBuffer` balance of `src` tokens.

Regarding expressing in terms of blocks - acknowledged. The preference if to stay with terms of seconds, as used for example in dss-flappers and in other places in the Maker ecosystem.